### PR TITLE
feat: Implement firmwareVersionCheck connect method to verify firmware revision

### PR DESCRIPTION
**Pull request by peter-sanderson**
Adds a check for firmware revision.

- If analytics are ON, we send this error to Sentry: 
image

- Check can be turned off the same way as Authenticity Check for newer devices
image

### CODE REVIEW

File: packages/connect/src/api/checkFirmwareAuthenticity.ts
Comment (peter-sanderson) [inline packages/connect/src/api/checkFirmwareAuthenticity.ts:61]: image

File: packages/connect-common/files/firmware/t1b1/releases.json
Comment (peter-sanderson) [inline packages/connect-common/files/firmware/t1b1/releases.json]: - [ ] TODO: DO NOT MERGE THIS

This has separate PR

File: packages/connect/src/device/checkFirmwareRevision.ts
Comment (peter-sanderson) [inline packages/connect/src/device/checkFirmwareRevision.ts]: - [x] Dont forget remove this

File: packages/suite/src/views/settings/SettingsDevice/SettingsDevice.tsx
Comment (peter-sanderson) [inline packages/suite/src/views/settings/SettingsDevice/SettingsDevice.tsx:91]: The `T3B1` is missing here (most likely by error). => Refactored to use the type-safe map so we know to add this param or new device next time.

File: packages/connect/src/types/device.ts
Comment (mroz22) [inline packages/connect/src/types/device.ts]: yes, it makes sense to me. define logic for firmwareHashCheck and authenticityCheck similarly to firmwareRevision check so that it is available also internally in connect(not only on API), and save its result value with device object. 

searching for volunteers 🧑‍🏭

File: packages/suite/src/constants/suite/device.ts
Comment (mroz22) [inline packages/suite/src/constants/suite/device.ts]: nitpick: just maybe naming could be better. isn't `SUPPORTS_DEVICE_AUTH`. too generic? meaning that there might be a device that supports firmware revision check but doesn't not support firmware hash checek.

File: packages/connect/src/device/checkFirmwareRevision.ts
Comment (mroz22) [inline packages/connect/src/device/checkFirmwareRevision.ts]: nitpick: maybe `if (!onlineRelease?.firmware_revision)`

File: packages/connect/src/device/checkFirmwareRevision.ts
Comment (mroz22) [inline packages/connect/src/device/checkFirmwareRevision.ts]: sooo, there is no local first variant?

File: packages/suite/src/components/suite/Preloader/Preloader.tsx
Comment (peter-sanderson) [inline packages/suite/src/components/suite/Preloader/Preloader.tsx:87]: - [x] Todo: send to Sentry/Analytics that this happened (check with Matěj, consult with Device Auth Check code how it does it)

File: packages/suite/src/components/suite/banners/SuiteBanners/SuiteBanners.tsx
Comment (MiroslavProchazka) [inline packages/suite/src/components/suite/banners/SuiteBanners/SuiteBanners.tsx:1]: We need to review those priorities against banners from config.v1.json, so they make sense altogether. Increasing the limit of priority above 100 is also on the table.

File: packages/connect/src/device/checkFirmwareRevision.ts
Comment (peter-sanderson) [reply packages/connect/src/device/checkFirmwareRevision.ts]: Yes there is. It is provided into the function through params from here:

image

File: packages/suite/src/constants/suite/device.ts
Comment (peter-sanderson) [reply packages/suite/src/constants/suite/device.ts]: https://github.com/trezor/trezor-suite/pull/13495/commits/8ff596a3791bd5e8ea78836d9ad10fed3135bd79

File: packages/connect/src/device/checkFirmwareRevision.ts
Comment (peter-sanderson) [reply packages/connect/src/device/checkFirmwareRevision.ts]: https://github.com/trezor/trezor-suite/pull/13495/commits/995f935c1a7bc3f779f2005997d28bfaa3a6e919

File: packages/suite/src/components/suite/Preloader/Preloader.tsx
Comment (szymonlesisz) [inline packages/suite/src/components/suite/Preloader/Preloader.tsx]: forgotten console.log

File: packages/connect/src/types/device.ts
Comment (szymonlesisz) [inline packages/connect/src/types/device.ts]: i would not step out from the convention `success: boolean` that we use everywhere else.

im not a big fan of using `null` as a meaningful value and distinguishing between `null` and `undefined` because after a while nobody remembers why it's done like that

i think this could be be done like:

```
{
   success: false,
   error: 'cannot-perform-check-offline'
}
```

and evaluated in SuiteBanners and suite Preloader using specific error literal

```
} else if (
    !isFirmwareRevisionCheckDisabled &&
    device?.authenticityChecks.firmwareRevision?.error === 'cannot-perform-check-offline' 
) {

```

File: packages/connect/src/device/Device.ts
Comment (mroz22) [inline packages/connect/src/device/Device.ts:552]: note: what happens if we run this on some very slow internet? maybe there should be time limit.

File: packages/connect/src/device/checkFirmwareRevision.ts
Comment (mroz22) [reply packages/connect/src/device/checkFirmwareRevision.ts]: optimiziation idea: 
maybe we could pull loading of the latest list of releases somewhere higher so that it doesn't block device features update 🤔

File: packages/connect/src/types/firmware.ts
Comment (szymonlesisz) [inline packages/connect/src/types/firmware.ts:20]: this is confusing,
i dont see `firmware_revision` field in either online (https://data.trezor.io/firmware/t1b1/releases.json) or offline (data dir) releases jsons.


does it mean it will fail here or i missed some deeper logic behind it?

```
if (onlineRelease?.firmware_revision === undefined) {
    return failFirmwareRevisionCheck('firmware-version-unknown');
}
```

File: packages/connect/src/device/checkFirmwareRevision.ts
Comment (szymonlesisz) [reply packages/connect/src/device/checkFirmwareRevision.ts]: plus we can download them once and then update current releases stored in DataManager (from what i understand it would fetch new releases on each device connect event)

File: packages/connect/src/device/__tests__/checkFirmwareRevision.test.ts
Comment (szymonlesisz) [inline packages/connect/src/device/__tests__/checkFirmwareRevision.test.ts]: nitpick: since this i only a mock, you dont need to put real changelog here

File: packages/connect/src/device/checkFirmwareRevision.ts
Comment (peter-sanderson) [reply packages/connect/src/device/checkFirmwareRevision.ts]: I like the idea, but this shall happen only rarely (newer firmware in older suite). I expect this to be a super-edge-case and its not worth it to make more complicated code just to address it. WDYT?

File: packages/connect/src/device/Device.ts
Comment (peter-sanderson) [reply packages/connect/src/device/Device.ts:552]: When I added:

```ts
await new Promise(r => setTimeout(r, 60 * 1000));
```

It got stucked on this for 60 seconds:
image

- Then it begins to do discovery...
- and after a second or two you get Ghosts.

File: packages/connect/src/device/checkFirmwareRevision.ts
Comment (komret) [inline packages/connect/src/device/checkFirmwareRevision.ts]: I think `firmwareVersion` would be more appropriate, it isn't immediately clear what `deviceVersion` is.

File: packages/connect/src/types/device.ts
Comment (peter-sanderson) [reply packages/connect/src/types/device.ts]: Good point :+1:

File: packages/connect/src/types/firmware.ts
Comment (peter-sanderson) [reply packages/connect/src/types/firmware.ts:20]: It will be there on data.trezor.io. @MiroslavProchazka shall finish his PR https://github.com/trezor/data/pull/88 :grin:

Comment (peter-sanderson) [reply packages/connect/src/types/firmware.ts:20]: If the FW release is not found locally AND its not found online it will show ghosts. 

So it is absolutely **crucial** to not forget to add `firmware_revision` to the data.trezor.io **before** we release new FW version.

File: packages/connect/src/types/device.ts
Comment (peter-sanderson) [reply packages/connect/src/types/device.ts]: https://github.com/trezor/trezor-suite/pull/13495/commits/7b2d5ae513b9c6c2be44bfeb57e4646ef54f8ae2

File: packages/connect/src/device/__tests__/checkFirmwareRevision.test.ts
Comment (peter-sanderson) [reply packages/connect/src/device/__tests__/checkFirmwareRevision.test.ts]: https://github.com/trezor/trezor-suite/pull/13495/commits/7b2d5ae513b9c6c2be44bfeb57e4646ef54f8ae2

File: packages/suite/src/components/suite/Preloader/Preloader.tsx
Comment (peter-sanderson) [reply packages/suite/src/components/suite/Preloader/Preloader.tsx]: https://github.com/trezor/trezor-suite/pull/13495/commits/7b2d5ae513b9c6c2be44bfeb57e4646ef54f8ae2

File: packages/suite/src/actions/suite/suiteActions.ts
Comment (komret) [inline packages/suite/src/actions/suite/suiteActions.ts]: I'd rename to verb form as we usually do for functions, i.e. `checkFirmwareRevision`.

File: packages/suite/src/components/suite/SecurityCheck/DeviceCompromised.tsx
Comment (komret) [inline packages/suite/src/components/suite/SecurityCheck/DeviceCompromised.tsx]: _Firmware_ revision check failed!

File: packages/suite/src/components/suite/SecurityCheck/DeviceCompromised.tsx
Comment (komret) [inline packages/suite/src/components/suite/SecurityCheck/DeviceCompromised.tsx]: The effect runs multiple times due to an object in the dependency array. The relevant properties should be dissected somehow.

File: packages/suite/src/components/suite/banners/SuiteBanners/SuiteBanners.tsx
Comment (komret) [inline packages/suite/src/components/suite/banners/SuiteBanners/SuiteBanners.tsx]: Isn't this condition redundant? What if Suite goes online in the instant between the check and this?

File: packages/suite/src/components/suite/modals/ReduxModal/UserContextModal/FirmwareRevisionCheckModal.tsx
Comment (komret) [inline packages/suite/src/components/suite/modals/ReduxModal/UserContextModal/FirmwareRevisionCheckModal.tsx:1]: I'd rename to `FirmwareRevisionOptOutModal` to match `DeviceAuhenticityOptOutModal`.

I also suggest creating a a separate component for shared logic between the two(plus possible `WipeDeviceModal` and some others. While at it, you could update to a more modern syntax (e.g. use `spacingsPx`) and clean up the styles (bold text everywhere looks lame, checkbox might look better aligned to the left).

File: packages/suite/src/components/suite/modals/ReduxModal/UserContextModal/UserContextModal.tsx
Comment (komret) [inline packages/suite/src/components/suite/modals/ReduxModal/UserContextModal/UserContextModal.tsx]:
```suggestion
        case 'firmware-revision-opt-out':
```

File: packages/suite/src/support/messages.ts
Comment (komret) [inline packages/suite/src/support/messages.ts]: Seems unintended.

File: packages/suite/src/support/messages.ts
Comment (komret) [inline packages/suite/src/support/messages.ts]: As above.

File: packages/connect/src/device/Device.ts
Comment (peter-sanderson) [reply packages/connect/src/device/Device.ts:552]: @mroz22 Finally I was able to test and simulate it.

image
image

With this setup:

image

File: packages/suite/src/components/suite/banners/SuiteBanners/SuiteBanners.tsx
Comment (peter-sanderson) [inline packages/suite/src/components/suite/banners/SuiteBanners/SuiteBanners.tsx:61]: Intentionally not checking the `!online` here. Downloading the `releases.json` may fail (blocked, filtered, ..) regardless of being online. 

For example, when testing it and filtering it in the networks tab :)

File: packages/connect/src/types/firmware.ts
Comment (szymonlesisz) [reply packages/connect/src/types/firmware.ts:20]: if its crucial then it should be covered by some pre-release e2e test

File: packages/connect/src/types/device.ts
Comment (mroz22) [inline packages/connect/src/types/device.ts]: btw isn't error: null kind of redundant?

File: packages/connect/src/device/checkFirmwareRevision.ts
Comment (mroz22) [reply packages/connect/src/device/checkFirmwareRevision.ts]: I think it is good enough now. it's an edgecase + it is timeoutable now, so even if it appears on the happy path (which it should not), it will not block it from execution too much.

File: packages/connect/src/types/firmware.ts
Comment (mroz22) [reply packages/connect/src/types/firmware.ts:20]: also this PR should get the update of releases.json before merging, right?

Comment (peter-sanderson) [reply packages/connect/src/types/firmware.ts:20]: @mroz22 already done https://github.com/trezor/trezor-suite/pull/13502

File: packages/connect/src/device/Device.ts
Comment (mroz22) [reply packages/connect/src/device/Device.ts:552]: fair enough

File: packages/suite/src/views/settings/SettingsDevice/SettingsDevice.tsx
Comment (komret) [inline packages/suite/src/views/settings/SettingsDevice/SettingsDevice.tsx]:
```suggestion
    const supportsFirmwareRevisionCheck = !supportsDeviceAuthentication; // Older devices with no secure element
```

Let's list the two devices to avoid some unexpected behaviour with new models.

File: packages/connect/src/types/device.ts
Comment (peter-sanderson) [reply packages/connect/src/types/device.ts]: https://github.com/trezor/trezor-suite/pull/13495/commits/e410e009138e31a4ed6414cf5cf3203547ea272c

**Review by komret — CHANGES_REQUESTED** [review]

File: packages/connect/src/device/checkFirmwareRevision.ts
Comment (peter-sanderson) [reply packages/connect/src/device/checkFirmwareRevision.ts]: [31c4b55371](https://github.com/trezor/trezor-suite/pull/13495/commits/31c4b553717a4b4bd26d067457df72236801cbd9)

File: packages/suite/src/actions/suite/suiteActions.ts
Comment (peter-sanderson) [reply packages/suite/src/actions/suite/suiteActions.ts]: https://github.com/trezor/trezor-suite/pull/13495/commits/3d2414e313716a93e704f1da746fb49491d2616d

File: packages/suite/src/components/suite/banners/SuiteBanners/SuiteBanners.tsx
Comment (peter-sanderson) [reply packages/suite/src/components/suite/banners/SuiteBanners/SuiteBanners.tsx]: Yup, see: https://github.com/trezor/trezor-suite/pull/13495#discussion_r1704204027

File: packages/suite/src/views/settings/SettingsDevice/SettingsDevice.tsx
Comment (peter-sanderson) [reply packages/suite/src/views/settings/SettingsDevice/SettingsDevice.tsx]: https://github.com/trezor/trezor-suite/pull/13495/commits/5b654e530778665277c72d2d093b910d4ccb91d8

File: packages/suite/src/support/messages.ts
Comment (peter-sanderson) [reply packages/suite/src/support/messages.ts]: https://github.com/trezor/trezor-suite/pull/13495/commits/5b654e530778665277c72d2d093b910d4ccb91d8

File: packages/suite/src/support/messages.ts
Comment (peter-sanderson) [reply packages/suite/src/support/messages.ts]: https://github.com/trezor/trezor-suite/pull/13495/commits/5b654e530778665277c72d2d093b910d4ccb91d8

File: packages/suite/src/components/suite/modals/ReduxModal/UserContextModal/UserContextModal.tsx
Comment (peter-sanderson) [reply packages/suite/src/components/suite/modals/ReduxModal/UserContextModal/UserContextModal.tsx]: https://github.com/trezor/trezor-suite/pull/13495/commits/5b654e530778665277c72d2d093b910d4ccb91d8

File: packages/suite/src/components/suite/modals/ReduxModal/UserContextModal/FirmwareRevisionCheckModal.tsx
Comment (peter-sanderson) [reply packages/suite/src/components/suite/modals/ReduxModal/UserContextModal/FirmwareRevisionCheckModal.tsx:1]: https://github.com/trezor/trezor-suite/pull/13495/commits/5b654e530778665277c72d2d093b910d4ccb91d8

File: packages/suite/src/components/suite/SecurityCheck/DeviceCompromised.tsx
Comment (peter-sanderson) [reply packages/suite/src/components/suite/SecurityCheck/DeviceCompromised.tsx]: https://github.com/trezor/trezor-suite/pull/13495/commits/5b654e530778665277c72d2d093b910d4ccb91d8

File: packages/connect/src/device/Device.ts
Comment (mroz22) [inline packages/connect/src/device/Device.ts:574]: I think this shouldn't actually happen at all. but I understand this check is needed because of typescript, right?

File: packages/connect/src/utils/assets.ts
Comment (mroz22) [inline packages/connect/src/utils/assets.ts:16]: side note: how is this change related with https://github.com/trezor/trezor-suite/issues/13271

**Review by mroz22 — APPROVED** [review]
connect part ack

File: packages/connect/src/device/Device.ts
Comment (peter-sanderson) [reply packages/connect/src/device/Device.ts:574]: image

**Review by komret — APPROVED** [review]

**Review by MiroslavProchazka — APPROVED** [review]

**Pull request merged**
